File: sample-chat-obj-c/sample-chat/Components/ChatScreen/SelectAssetsScreen/SelectAssetsVC.m
//
//  SelectAssetsVC.m
//  sample-chat
//
//  Created by Injoit on 2/6/20.
//  Copyright © 2020 Quickblox. All rights reserved.
//

#import "SelectAssetsVC.h"
#import <Photos/Photos.h>
#import "UIView+Chat.h"
#import "SelectAssetCell.h"
#import <math.h>
#import "UIViewController+Alert.h"
#import "UICollectionView+Chat.h"
#import "UIViewController+Alert.h"

const CGFloat itemsInRow = 3.0f;
const double kMaximumMB = 90;
const double kDividerToMB = 1048576;
const CGFloat kMinimumSpacing = 8.0f;
NSString *const MAX_FILE_SIZE = @"The uploaded file exceeds maximum file size (100MB)";
NSString *const ERROR_IMAGE_LOAD = @"Error loading image";

@interface SelectAssetsVC () <UICollectionViewDelegate,
UICollectionViewDataSource, UICollectionViewDelegateFlowLayout, PHPhotoLibraryChangeObserver>
#pragma mark - Properties
@property (weak, nonatomic) IBOutlet UICollectionView *collectionView;
@property (weak, nonatomic) IBOutlet UIButton *cancelButton;
@property (weak, nonatomic) IBOutlet UIView *containerView;
@property (weak, nonatomic) IBOutlet UIButton *sendAttachmentButton;
@property (strong, nonatomic) PHAsset *selectedAsset;
@property (strong, nonatomic) PHFetchResult<id> *allPhotos;
@property (strong, nonatomic) PHCachingImageManager *cachingImageManager;
@property (assign, nonatomic) CGRect previousRect;
@property (assign, nonatomic) CGSize thumbnailImageSize;
@end

@implementation SelectAssetsVC
#pragma mark - Life Cycle
- (void)viewDidLoad {
    [super viewDidLoad];

    [PHPhotoLibrary.sharedPhotoLibrary registerChangeObserver:self];
    
    self.cachingImageManager = [[PHCachingImageManager alloc] init];
    [self reloadImages];
    
    [self.containerView roundTopCornersWithRadius:14.0f];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    CGFloat width = UIScreen.mainScreen.bounds.size.width/itemsInRow - kMinimumSpacing * (itemsInRow - 1);
    self.thumbnailImageSize = CGSizeMake(width, width);
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    
    [self updateCachedImages];
}

#pragma mark - PHPhotoLibraryChangeObserver
- (void)photoLibraryDidChange:(PHChange *)changeInstance {
    __weak typeof(self)weakSelf = self;
    dispatch_async(dispatch_get_main_queue(), ^{
        [weakSelf reloadImages];
        [weakSelf.collectionView reloadData];
        [weakSelf updateCachedImages];
    });
}

- (void)reloadImages {
    [self stopCachingImages];
    PHFetchOptions *fetchOptions = [[PHFetchOptions alloc] init];
    fetchOptions.includeAllBurstAssets = YES;
    fetchOptions.sortDescriptors = @[[NSSortDescriptor sortDescriptorWithKey:@"creationDate" ascending:YES]];
    self.allPhotos = [PHAsset fetchAssetsWithOptions:fetchOptions];
}

#pragma mark - Internal Methods
- (void)stopCachingImages {
    [self.cachingImageManager stopCachingImagesForAllAssets];
    self.previousRect = CGRectZero;
}

- (void)updateCachedImages {
    BOOL isViewVisible = [self isViewLoaded] && [[self view] window] != nil;
    if (!isViewVisible) { return; }

    CGRect originalRect = self.collectionView.bounds;
    originalRect = CGRectInset(originalRect, 0.0f, -0.5f * CGRectGetHeight(originalRect));

    CGFloat deltaY = ABS(CGRectGetMidY(originalRect) - CGRectGetMidY(self.previousRect));
    if (deltaY > CGRectGetHeight(originalRect) / 3.0f) {

        NSMutableArray *addedIndexPaths = [NSMutableArray array];
        NSMutableArray *removedIndexPaths = [NSMutableArray array];
        
        [self.collectionView compareHandlerWithRect:self.previousRect andRect:originalRect removedHandler:^(CGRect removedRect) {
            NSArray *indexPaths = [self.collectionView indexPathsForElementsInRect:removedRect];
            [removedIndexPaths addObjectsFromArray:indexPaths];
        } addedHandler:^(CGRect addedRect) {
            NSArray *indexPaths = [self.collectionView indexPathsForElementsInRect:addedRect];
            [addedIndexPaths addObjectsFromArray:indexPaths];
        }];
        
        NSArray *assetsStartCaching = [self assetsAtIndexPaths:addedIndexPaths];
        NSArray *assetsStopCaching = [self assetsAtIndexPaths:removedIndexPaths];
        
        [self.cachingImageManager startCachingImagesForAssets:assetsStartCaching
                                            targetSize:self.thumbnailImageSize
                                           contentMode:PHImageContentModeAspectFill
                                               options:nil];
        [self.cachingImageManager stopCachingImagesForAssets:assetsStopCaching
                                           targetSize:self.thumbnailImageSize
                                          contentMode:PHImageContentModeAspectFill
                                              options:nil];
        
        self.previousRect = originalRect;
    }
}

- (NSArray *)assetsAtIndexPaths:(NSArray *)indexPaths {
    if (indexPaths.count == 0) { return nil; }
    
    NSMutableArray *assets = [NSMutableArray arrayWithCapacity:indexPaths.count];
    for (NSIndexPath *indexPath in indexPaths) {
        PHAsset *asset = self.allPhotos[indexPath.item];
        [assets addObject:asset];
    }
    return assets;
}

#pragma mark - Actions
- (IBAction)cancelButtonTapped:(UIButton *)sender {
    [self dismissViewControllerAnimated:YES completion:^{
        if (self.selectedImage) {
            self.selectedImage(nil);
        }
    }];
}

- (IBAction)sendAttachmentButtonTapped:(UIButton *)sender {
    if (!self.selectedAsset) {return;}
    void(^sendAsset)(double size) = ^(double size) {
        double sizeMB = size/kDividerToMB;
        sizeMB = trunc(sizeMB * 100) / 100;
        if (sizeMB > kMaximumMB) {
            [self showAlertWithTitle:MAX_FILE_SIZE message:nil
                  fromViewController:self handler:nil];
        } else {
            self.sendAttachmentButton.enabled = NO;
            PHImageRequestOptions *options = [[PHImageRequestOptions alloc] init];
            options.deliveryMode = PHImageRequestOptionsDeliveryModeHighQualityFormat;
            options.networkAccessAllowed = YES;
            
            __weak typeof(self)weakSelf = self;
            [PHImageManager.defaultManager requestImageForAsset:self.selectedAsset targetSize:PHImageManagerMaximumSize contentMode:PHImageContentModeAspectFill options:options resultHandler:^(UIImage * _Nullable resultImage, NSDictionary * _Nullable info) {
                __typeof(weakSelf)strongSelf = weakSelf;
                if (resultImage) {
                    dispatch_async(dispatch_get_main_queue(), ^{
                        [strongSelf dismissViewControllerAnimated:YES completion:^{
                            if (strongSelf.selectedImage) {
                                strongSelf.selectedImage(resultImage);
                            }
                        }];
                    });
                } else {
                    [strongSelf showAlertWithTitle:ERROR_IMAGE_LOAD message:nil
                                fromViewController:strongSelf handler:^(UIAlertAction * _Nonnull action) {
                        strongSelf.sendAttachmentButton.enabled = NO;
                    }];
                }
            }];
        }
    };
    
    [self.selectedAsset requestContentEditingInputWithOptions:nil completionHandler:^(PHContentEditingInput * _Nullable contentEditingInput, NSDictionary * _Nonnull info) {
        NSNumber *fileSize = nil;
        NSError *error = nil;
        [contentEditingInput.fullSizeImageURL getResourceValue:&fileSize forKey:NSURLFileSizeKey error:&error];
        sendAsset(fileSize.doubleValue);
    }];
    
}

#pragma mark - UICollectionViewDelegate, UICollectionViewDataSource
- (NSInteger)collectionView:(nonnull UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return self.allPhotos.count;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    PHAsset *phAsset = [self.allPhotos objectAtIndex:indexPath.item];
    
    SelectAssetCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"SelectAssetCell" forIndexPath:indexPath];
    cell.representedAssetIdentifier = phAsset.localIdentifier;
    [self.cachingImageManager requestImageForAsset:phAsset
                                        targetSize:self.thumbnailImageSize
                                       contentMode:PHImageContentModeAspectFill
                                           options:nil
                                     resultHandler:^(UIImage * _Nullable result, NSDictionary * _Nullable info) {
        if ([cell.representedAssetIdentifier isEqualToString:phAsset.localIdentifier]) {
            cell.assetImageView.image = result;
        }
    }];
    return cell;
}

-(void)scrollViewDidScroll:(UIScrollView *)scrollView {
    [self updateCachedImages];
}

- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath {
    self.selectedAsset = [self.allPhotos objectAtIndex:indexPath.item];
    self.sendAttachmentButton.hidden = !collectionView.indexPathsForSelectedItems.count;
}

#pragma mark - UICollectionViewDelegateFlowLayout
- (CGSize)collectionView:(UICollectionView *)collectionView
                  layout:(UICollectionViewLayout *)collectionViewLayout
  sizeForItemAtIndexPath:(NSIndexPath *)indexPath {
    CGFloat width = UIScreen.mainScreen.bounds.size.width/itemsInRow - kMinimumSpacing * (itemsInRow - 1);
    return CGSizeMake(width, width);
}

- (CGFloat)collectionView:(UICollectionView *)collectionView
                   layout:(UICollectionViewLayout *)collectionViewLayout
minimumLineSpacingForSectionAtIndex:(NSInteger)section {
    return kMinimumSpacing;
}

- (CGFloat)collectionView:(UICollectionView *)collectionView
                   layout:(UICollectionViewLayout *)collectionViewLayout
minimumInteritemSpacingForSectionAtIndex:(NSInteger)section {
    return kMinimumSpacing;
}

@end

File: sample-conference-videochat-swift/sample-conference-videochat-swift/Components/ChatScreen/SelectAssetsScreen/SelectAssetsVC.swift
//
//  SelectAssetsVC.swift
//  sample-conference-videochat-swift
//
//  Created by Injoit on 12/9/19.
//  Copyright © 2019 Quickblox. All rights reserved.
//

import UIKit
import Photos
import PDFKit
import SVProgressHUD

struct PhotoAsset: Equatable {
    var phAsset: PHAsset
    var image: UIImage
}

struct SelectAssetsConstant {
    static let maximumMB: Double = 100
    static let dividerToMB: Double = 1048576
    static let minimumSpacing: CGFloat = 8.0
    static let reuseIdentifier = "SelectAssetCell"
}

class SelectAssetsVC: UIViewController {
    @IBOutlet weak var cancelButton: UIButton!
    @IBOutlet weak var containerView: UIView!
    @IBOutlet weak var sendAttachmentButton: UIButton!
    @IBOutlet weak var collectionView: UICollectionView!
    fileprivate var allPhotos: PHFetchResult<AnyObject>?
    var selectedAssetCompletion:((_ asset: PhotoAsset?) -> Void)?
    var selectedAsset: PhotoAsset?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        fetchAssets()
        containerView.roundTopCorners()
    }
    
    private func fetchAssets() {
        let fetchOptions = PHFetchOptions()
        fetchOptions.includeAssetSourceTypes = [.typeUserLibrary]
        fetchOptions.sortDescriptors = [NSSortDescriptor.init(key: "creationDate", ascending: false)]
        allPhotos = nil
        collectionView.reloadData()
        if let allPhotos = PHAsset.fetchAssets(with: PHAssetMediaType.image, options: fetchOptions) as? PHFetchResult<AnyObject> {
            self.allPhotos = allPhotos
            self.collectionView.reloadData()
            SVProgressHUD.dismiss()
        }
    }
    
    @IBAction func cancelButtonTapped(_ sender: UIButton) {
        self.dismiss(animated: true) {
            self.selectedAssetCompletion?(nil)
        }
    }
    
    @IBAction func sendAttachmentButtonTapped(_ sender: UIButton) {
        let sendAsset: (Double) -> Void = { [weak self] (size) in
            let sizeMB = size/SelectAssetsConstant.dividerToMB
            if sizeMB.truncate(to: 2) > SelectAssetsConstant.maximumMB {
                self?.showAlertView("The uploaded file exceeds maximum file size (100MB)", message: nil)
            } else {
                SVProgressHUD.show()
                self?.sendAttachmentButton.isEnabled = false
                if let selectedAsset = self?.selectedAsset {
                    let options = PHImageRequestOptions()
                    options.deliveryMode = .highQualityFormat
                    PHImageManager.default().requestImage(for: selectedAsset.phAsset, targetSize: PHImageManagerMaximumSize, contentMode: .aspectFill, options: options) { (image, info) -> Void in
                        SVProgressHUD.dismiss()
                        if let image = image {
                            let photoAsset = PhotoAsset(phAsset: selectedAsset.phAsset, image: image)
                            DispatchQueue.main.async {
                                self?.dismiss(animated: true) {
                                    self?.selectedAssetCompletion?(photoAsset)
                                }
                            }
                        } else {
                            self?.showAlertView("Error loading image", message: nil)
                        }
                    }
                }
            }
        }
        if let selectedAsset = selectedAsset {
            let res = PHAssetResource.assetResources(for: selectedAsset.phAsset)
            if let assetSize = res.first?.value(forKey: "fileSize") as? Double {
                sendAsset(assetSize)
            }
        }
    }
}

extension SelectAssetsVC: UICollectionViewDelegate, UICollectionViewDataSource {
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        if let allPhotos = allPhotos {
            return allPhotos.count
        }
        return 0
    }
    
    func collectionView(_ collectionView: UICollectionView, willDisplay cell: UICollectionViewCell, forItemAt indexPath: IndexPath) {
        guard let cell = cell as? SelectAssetCell, let phAsset = allPhotos?[indexPath.row] as? PHAsset, let size = cell.assetImageView?.bounds.size else {
            return
        }

        PHImageManager.default().requestImage(for: phAsset, targetSize: size, contentMode: .aspectFill, options: nil) { (image, info) -> Void in
            if let image = image {
                cell.assetImageView.image = image
            }
        }
    }

    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: SelectAssetsConstant.reuseIdentifier, for: indexPath) as? SelectAssetCell else {
            return UICollectionViewCell()
        }

        return cell
    }
    
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        guard let phAsset = allPhotos?[indexPath.row] as? PHAsset else {
            return
        }
        let photoAsset = PhotoAsset(phAsset: phAsset, image: UIImage())
        self.selectedAsset = photoAsset
        self.sendAttachmentButton.isHidden = collectionView.indexPathsForSelectedItems?.isEmpty == false ? false : true
    }
}

extension SelectAssetsVC: UICollectionViewDelegateFlowLayout {
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
        let itemsInRow: CGFloat = 3
        let width = (UIScreen.main.bounds.width)/itemsInRow - SelectAssetsConstant.minimumSpacing * (itemsInRow - 1)
        return CGSize(width: width, height: width)
    }

    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, minimumLineSpacingForSectionAt section: Int) -> CGFloat {
        return SelectAssetsConstant.minimumSpacing
    }

    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, minimumInteritemSpacingForSectionAt section: Int) -> CGFloat {
        return SelectAssetsConstant.minimumSpacing
    }
}
